nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-ssl-autoloading-utils/src/main/java/org/apache/nifi/autoload/SSLContextFactoryAutoLoaderTask.java

```diff
@@ -0,0 +1,270 @@
+            }
+        }
+    }
+
+    /**
+     * Returns a list representing the state of the current keystore at the given path.  This method uses the
+     * keystore properties from nifi.properties.  The only state retrieved will be the alias, subject DN,
+     * issuer DN of each PrivateKeyEntry, and issuer certificate serial number if applicable, and the results
+     * will be a sorted list.
+     * @return A sorted list of information about each private key in the keystore
+     * @throws TlsException If the keystore could not be loaded
+     * @throws KeyStoreException If the keystore password was incorrect
+     * @throws UnrecoverableEntryException If a private key entry could not be recovered
+     * @throws NoSuchAlgorithmException If the default password algorithm is not supported
+     */
+    private List<PrivateKeyInfo> getKeystoreState() throws TlsException, KeyStoreException,
+            UnrecoverableEntryException, NoSuchAlgorithmException {
+        List<PrivateKeyInfo> state = new ArrayList<>();
+
+        KeyStore keyStore = KeyStoreUtils.loadKeyStore(keystorePath.toString(),
+                nifiProperties.getProperty(NiFiProperties.SECURITY_KEYSTORE_PASSWD).toCharArray(),
+                nifiProperties.getProperty(NiFiProperties.SECURITY_KEYSTORE_TYPE));
+
+        Enumeration<String> aliasesEnum = keyStore.aliases();
+        while(aliasesEnum.hasMoreElements()) {
+            String alias = aliasesEnum.nextElement();
+            if (keyStore.isKeyEntry(alias)) {
+                KeyStore.PrivateKeyEntry entry = (KeyStore.PrivateKeyEntry) keyStore.getEntry(alias, new KeyStore.PasswordProtection(nifiProperties
+                        .getProperty(NiFiProperties.SECURITY_KEY_PASSWD).toCharArray()));
+                X509Certificate cert = (X509Certificate) entry.getCertificateChain()[0];
+                String issuerSerialNumber = (entry.getCertificateChain().length > 1)
+                        ? ((X509Certificate) entry.getCertificateChain()[1]).getSerialNumber().toString()
+                        : null;
+                state.add(new PrivateKeyInfo(cert.getSubjectDN().getName(), cert.getIssuerDN().getName(), alias, issuerSerialNumber));
+            }
+        }
+
+        Collections.sort(state);
+        return new ArrayList<>(state);
+    }
+
+    /**
+     * This returns false if there were any changes to the keystore other than updating a PrivateKeyEntry with
+     * the same subject DN, issuer DN, alias, and issuer cert serial number if applicable.
+     * @return True if a reload should be allowed, meaning the keystore has not changed in a meaningful way
+     */
+    boolean isReloadAllowed() throws NoSuchAlgorithmException, UnrecoverableEntryException, KeyStoreException, TlsException {
+        return existingKeystoreState.equals(this.getKeystoreState());
+    }
+
+    public void stop() {
+        LOGGER.info("Stopping SSL Context Factory Auto-loader");
+        stopped = true;
+    }
```

Comment:
Rather than having a `Runnable` with its own lifecycle method, what about leveraging `ScheduledThreadPoolExecutor.scheduleAtFixedRate()` and managing the lifecycle through the returned `ScheduledFuture`?  That would make it much easier to test this class without handling threading concerns.

Reply:
I like it, and this enabled me to write some additional unit tests.